ynr/apps/official_documents/static/official_documents/js/sopn_viewer.js

```diff
@@ -0,0 +1,96 @@
+            canvas.width = viewport.width;
+            let renderContext = {
+                canvasContext: context,
+                viewport: viewport
+            };
+
+            let renderTask = page.render(renderContext);
+            renderTask.then(function () {
+                window.pdf_container.append(page_container);
+                return page.getTextContent({normalizeWhitespace: true});
+            }).then(function (textContent) {
+                let pdf_canvas = $(canvas);
+                let page_text_layer = document.createElement("div");
+                page_text_layer.className = "text_layer";
+                page_container.appendChild(page_text_layer);
+
+                let canvas_offset = pdf_canvas.offset();
+                let canvas_height = pdf_canvas.get(0).height;
+                let canvas_width = pdf_canvas.get(0).width;
+
+                $(page_text_layer).css({
+                    left: canvas_offset.left + 'px',
+                    top: canvas_offset.top + 'px',
+                    height: canvas_height + 'px',
+                    width: canvas_width + 'px'
+                });
+
+                pdfjsLib.renderTextLayer({
+                    textContent: textContent,
+                    container: page_text_layer,
+                    viewport: viewport,
+                    textDivs: []
+                });
+                // Load more pages when this page is done
+                start_page++;
+                load_pages_by_range(pdf, start_page, end_page);
+            });
+
+        }
+
+    });
+}
+
+function ShowSOPNInline(sopn_url, ballot_paper_id, options) {
+    const loadingTask = pdfjsLib.getDocument(sopn_url);
+
+    loadingTask.promise.then(function(pdf) {
+        // The container element
+        window.pdf_container = document.getElementById('sopn-' + ballot_paper_id);
+
+        // Get the end page either from the defined range, or the total number of pages
+        const start_page = options.start_page || 1;
+        const end_page = options.end_page || pdf.numPages;
+
+        load_pages_by_range(pdf, start_page, end_page);
+
+    }).then(null, function(error) {
+        window.pdf_container = document.getElementById('sopn-' + ballot_paper_id);
+        if (error.name === "MissingPDFException") {
+            window.pdf_container.innerHTML = "<h3>PDF file not found</h3>"
+        }
+
+        if (error.name === "InvalidPDFException") {
+            /*
+            Show the Google document viewer if PDFJS can't deal with this document for whatever reason
+            */
+            google_frame = document.createElement("iframe");
+            google_frame.setAttribute("frameborder", 0);
+            google_frame.setAttribute("allowfullscreen", true);
+            google_frame.className = "document_viewer";
+            url = "https://docs.google.com/viewer?url=https://candidates.democracyclub.org.uk"
+                +encodeURI(sopn_url)+"&amp;embedded=true";
```

Comment:
`+` on a continuation line may lead to ASI bugs in some environments. End the previous line with a `+` to split this statement over >1 line

Reply:
..also shouldn't this be `&embedded=true` not `&amp;embedded=true`. Remember this is a `.js` file not a `.html` file

Out of interest, how did you manage to test this locally? I get `Load denied by X-Frame-Options docs.google.com does not permit cross-origin framing`, trying to test it.. which makes sense, but I'm just wondering..

Reply:
Tested by just checking the URL produced was the same as written in to the HTML of the previous implementation. Oddly we were using `&amp;` before and it didn't change anything. Not sure what the deal is there – maybe `embedded` doesn't actually do anything?

Here's an example URL to have a look at:

https://docs.google.com/viewer?url=https://www3.northamptonshire.gov.uk/councilservices/council-and-democracy/councillors-and-committees/Documents/Oundle_Statement_of_Persons_Nominated.pdf&embedded=true

Reply:
`&embedded=true` seems to hide the menu with a print button and stuff for me, but it makes relatively little difference. Maybe this was always "broken", but it didn't really matter :)